src/node/txreconciliation.cpp

```diff
@@ -193,6 +210,104 @@ class TxReconciliationTracker::Impl
         LOCK(m_txreconciliation_mutex);
         return IsPeerRegistered(peer_id);
     }
+
+    std::vector<NodeId> GetFanoutTargets(CSipHasher& deterministic_randomizer_with_wtxid,
```

Comment:
I wonder if this algorithm (which took me a while to fully understand) could be simpler. 
E.g., if we used a sorted container for `best_peers` instead of a vector, inserted all of the peers, and then finally return the first `targets` elements of that container, I think we could do without the `try_fanout_candidate` lambda.
Or would that be incorrect / less performant?

I'm thinking of something like the following (just to show idea, I didn't test it):

    struct ComparePairs {
        bool operator()(const std::pair<uint64_t, NodeId>& left, const std::pair<uint64_t, NodeId>& right) const {
            return left.first > right.first;
        }
    };
    std::vector<NodeId> GetFanoutTargets(CSipHasher& deterministic_randomizer_with_wtxid,
                                         bool we_initiate, double limit) const EXCLUSIVE_LOCKS_REQUIRED(m_txreconciliation_mutex)
    {
        // The algorithm works as follows. We iterate through the peers (of a given direction)
        // hashing them with the given wtxid, and sort them by this hash.
        // We then consider top `limit` peers to be low-fanout flood targets.
        // The randomness should be seeded with wtxid to return consistent results for every call.

        double integer_part;
        double fractional_peer = std::modf(limit, &integer_part);
        const bool drop_peer_if_extra = deterministic_randomizer_with_wtxid.Finalize() > fractional_peer * double(UINT64_MAX);
        const size_t targets = drop_peer_if_extra ? size_t(integer_part): size_t(integer_part) + 1;

        std::set<std::pair<uint64_t, NodeId>, ComparePairs> best_peers;

        for (auto indexed_state : m_states) {
            const auto cur_state = std::get_if<TxReconciliationState>(&indexed_state.second);
            if (cur_state && cur_state->m_we_initiate == we_initiate) {
                uint64_t hash_key = deterministic_randomizer_with_wtxid.Write(cur_state->m_k0).Finalize();
                best_peers.insert(std::make_pair(hash_key, indexed_state.first));
            }
        }

        std::vector<NodeId> result;
        auto it = best_peers.begin();
        for (size_t i = 0; i < targets && it != best_peers.end(); ++i, ++it) {
            result.push_back(it->second);
        }
        return result;
    }

Reply:
This is certainly better. For no good reason, i just chose to follow a pattern we use elsewhere and never reconsidered it. I will take this code.